Add ClientErrorCollector Helper

kitsonk opened the pull request:
This PR adds `ClientErrorCollector` to the kitchen sink repo.

`ClientErrorCollector` is a LeadFoot helper that is designed to capture client error in functional tests that might occur, and are not desired, but do not cause the functional test to fail.  It is a class that provides three methods:

- `init()` - Initialises the collector, installing a global `onerror` handler which will collect any errors
- `finish()` - Uninstall the collector and return a Promise that resolves to an array of any errors
- `assertNoErrors()` - Implicitly calls `finish()` and will reject if there are any errors, rejecting on the first error in the stack

This then is integrated into the functional tests for kitchen sink.

Refs #196

agubler commented:
Is this going to be added to Intern? Or do you think we should move the support module to either grunt-dojo2 or a dedicated test support package? As soon as we want this for todo-mvc or our Dojo 2 applications we'd need to copy the module around.

kitsonk commented:
@jason0x43 doesn't want it in Intern right now.

While Grunt Dojo 2 is currently a dependency for kitchensink, is it the right place for something like this?

matt-gadd commented on todo-mvc-kitchensink/tests/functional/Page.ts:
```diff
@@ -79,7 +83,10 @@ export default class Page {
-			.setFindTimeout(100);
+			.setFindTimeout(100)
+			.then(() => {
+				return this.errorCollector.init()
```
missing semi-colon. should've been picked up by lint?

kitsonk replied:
It did in the IDE, but it appears that we don't lint on `dojo build` so I missed the little green squiggle... Is that what we expect/want from `dojo build`?  Should it have an option to do linting?

dylans replied:
we have a cli linting repo but I don’t think we’ve done anything with it yet, https://github.com/dojo/cli-lint-tslint

matt-gadd commented on todo-mvc-kitchensink/tests/support/ClientErrorCollector.ts:
```diff
@@ -0,0 +1,115 @@
+			});
+	}
+
+		init(): Promise<void> {
```
indentation.

matt-gadd commented on todo-mvc-kitchensink/tests/support/ClientErrorCollector.ts:
```diff
@@ -0,0 +1,115 @@
+
+		return this._remote
+			.execute(`
+				var errorStack = window.${CLIENT_ERROR_STACK_KEY} = [];
```
Really not a fan of code as a string. You can pass a function and args to `.execute`. Making this a real function will catch any syntax errors, allow you to type it, and not have to write es5 code.

kitsonk replied:
I wasn't entirely sure I trusted `.toString()` on the function, but ok, yeah that makes sense.

matt-gadd commented on todo-mvc-kitchensink/tests/support/ClientErrorCollector.ts:
```diff
@@ -0,0 +1,115 @@
+					errorStack.push(errorObj);
+
+					if (oldonerror) {
+						oldonerror.call(this, messageOrEvent, source, lineno, colno, error);
```
where does `messageOrEvent` come from?

kitsonk replied:
Good catch... and valid to your earlier point! 😢

matt-gadd commented on todo-mvc-kitchensink/tests/support/ClientErrorCollector.ts:
```diff
@@ -0,0 +1,115 @@
+					const result = results[0];
+					let e: Error;
+					switch (result.type) {
+					case 'TypeError':
```
How come just `Type` and `Syntax` errors? (and not `Range` and `Reference`?).

kitsonk replied:
Fair point.

matt-gadd commented:
@agubler how come grunt-dojo2 is still used in kitchensink? I don't think we should be putting anything that we intend an app to use in grunt-dojo2.

agubler commented:
I'm not even sure it does, presuming we'd want this for internal functionals also. but regardless it credence to the need for a dedicated repository with support code such as this.

matt-gadd commented:
@agubler Yeah I think we should just create an intern-support repo or similar. We still have a number of duplicated test support files across repos. It's probably where we should have put the custom reporter also.

agubler commented:
@matt-gadd agreed.

kitsonk commented:
`dojo/intern-support`, then?

dylans commented:
> dojo/intern-support, then?

regarding the name intern-support, perhaps something more general like test-helpers or test-support or test-intern-helpers… I don’t know, just thinking of how we have cli-test-intern and maybe want to follow that pattern of intent and then name, but maybe that’s just for cli

matt-gadd commented:
@dylans I do think we should have the word `intern` somewhere in there, because this and the Reporter are both intern specific. Either that or we have a generic `support` repo with an `intern` folder, although I'd hope that wouldn't become a general dumping ground.

agubler commented:
@matt-gadd is the mock modules module intern specific? It's another thing I can think of that could go in this support repo.

matt-gadd commented:
@agubler indeed, it certainly has `intern` specific parts currently anyway.

kitsonk commented:
Also we have a couple other helpers scattered about too, like [Reporter](https://github.com/dojo/dojo2-package-template/blob/master/tests/support/Reporter.ts) and [util](https://github.com/dojo/dojo2-package-template/blob/master/tests/support/util.ts).

kitsonk commented:
I addressed most of the feedback but will break the functional code out to a separate package and then this PR can morph into implementing that package.

dylans commented:
I believe this has been addressed via the creation of https://github.com/dojo/test-extras/ .

kitsonk commented:
While this PR is invalid, since it was moved to `test-extras` we still need to integrate the `ClientErrorReporter`.
